### ISSUE TITLE
Handle multiple assignment to same allocatable variable

### ISSUE THREAD

**Issue by yguclu** [pyccel/pyccel#359]
In Python every assignment `x = expr` declares a new variable `x`, to which it assigns the value of `expr`.

When we generate Fortran code we map this Python expression to a Fortran assignment, and if necessary we precede it with the declaration of `x` (if it was never used before) and with a memory allocation (if `expr` is a Numpy array constructor and `x` was not declared with `@stack_array('x')`).

In all cases we cannot allow the user to write `x = expr` (i.e. we raise a PyccelError) if the two sides of the assignment have either

1. different `dtype`
2. different `rank`

Moreover, even if `x` and `expr` have same `dtype` and `rank`, the case where the value of `expr` is a Numpy array requires additional care. For the situations that require memory reallocation we raise a PyccelWarning:

3. if Numpy array `x` already exists
4. if Numpy array `x` is created inside of a loop
5. if Numpy array `x` is created inside of an if/else statement

For these situations we will add some Fortran code that checks if `x` was already allocated, and if the runtime shape of `x` and `expr` is different. Memory reallocation will only happen if needed, but Pyccel will not be able to predict that (hence the PyccelWarning).

In conclusion we need 3 new AST node classes:
1. AllocateNode
2. DeallocateNode (already exists under name of "Delete")
3. CheckReallocationNode (which uses nodes 1 and 2)

For now in semantic we will create a node of type 3, which then creates nodes of types 1 and 2 at the printing stage.

**Comment by yguclu** [comment]
Let's try to define a concrete example. We would like to translate the Python program
```
import numpy as np

i = 5
j = 7

a = np.ones (i, dtype=int)
b = np.zeros(j, dtype=int)

# Raises an error: incompatible redefinition
# file pyccel/parser/semantic.py, line 1466
b = 3 * a
```
to the Fortran program
```
program prog_ex_program

implicit none

integer(kind=8) :: i
integer(kind=8) :: j
integer(kind=8), allocatable :: a(:)
integer(kind=8), allocatable :: b(:)

i = 5_8
j = 7_8
allocate(a(0:i - 1_8))
a(:) = 1
allocate(b(0:j - 1_8))
b(:) = 0

! Raises an error: incompatible redefinition
! file pyccel/parser/semantic.py, line 1466
if (any(shape(b) /= shape(a))) then
    print *, "PYCCEL RUNTIME WARNING: array b is being reallocated"
    deallocate(b)
    allocate(b(0:size(a,1)-1))
end if
b(:) = 3 * a

end program prog_ex_program
```
In addition, Pyccel should raise a warning during the translation stage:
```
$  >> PYCCEL WARNING: redefinition of array may cause reallocation at runtime
```

**Comment by saidctb** [comment]
I think you are conflicting two behaviour, the first is multiple allocation and the second is automatic allocation
we should handle the first before going to the second because we could have something like

```
```

**Comment by yguclu** [comment]
@said The latter example that you present is already handled by pyccel:
```
import numpy as np

a = np.ones(5)
b = 3 * a
```
becomes
```
program prog_array_from_expression

implicit none

real(kind=8), allocatable :: a(:)
real(kind=8), allocatable :: b(:)

allocate(a(0:5_8 - 1_8))
a = 1.0
allocate(b(0:5_8 - 1_8))
b = 3_8 * a

end program prog_array_from_expression
```

**Comment by EmilyBourne** [comment]
So following the logic above, Said's first example should generate:
```
program prog_ex_program

implicit none

integer(kind=8) :: i
integer(kind=8) :: j
integer(kind=8), allocatable :: a(:)

i = 5_8
j = 7_8
allocate(a(0:i - 1_8))
a(:) = 1

! Raises an error: incompatible redefinition
! file pyccel/parser/semantic.py, line 1466
if (any(shape(a) /= ( j ))) then
    print *, "PYCCEL RUNTIME WARNING: array b is being reallocated"
    deallocate(a)
    allocate(a(0:j-1))
end if
a(:) = 0

end program prog_ex_program
```

**Comment by EmilyBourne** [comment]
I agree that we should concentrate on the simpler case:
```
a = np.ones (i, dtype=int)
a = np.zeros(j, dtype=int)
```

As the second case is already handled, there's a chance that fixing this case will automatically fix @yguclu 's slightly more complicated case

**Comment by yguclu** [comment]
First minimal and complete example that we want handled: program
```
import numpy as np

i = 5
j = 7

a = np.ones (i, dtype=int)
a = np.zeros(j, dtype=int)
```
should be translated to
```
program prog_array_redefinition

implicit none

integer(kind=8) :: i
integer(kind=8) :: j
integer(kind=8), allocatable :: a(:)

i = 5_8
j = 7_8
allocate(a(0:i - 1_8))
a(:) = 1

if (any(shape(a) /= [j])) then
    print *, "PYCCEL RUNTIME WARNING: array a is being reallocated"
    deallocate(a)
    allocate(a(0:j-1))
end if
a(:) = 0

end program prog_array_redefinition
```

**Comment by yguclu** [comment]
Second minimal example: **array creation in loop**. Program
```python
import numpy as np

n = 10
for i in range(n):
    a = np.ones(i)
```
should be translated to
```fortran
program prog_array_creation_in_loop

implicit none

integer(kind=8) :: n
integer(kind=8) :: i
real(kind=8), allocatable :: a(:)

n = 10_8
do i = 0_8, n-1_8, 1_8
  if (allocated(a) .and. any(shape(a) /= [i])) then
    deallocate(a)
  end if
  allocate(a(0:i - 1_8))
  a = 1.0

end do

end program prog_array_creation_in_loop
```
Currently Pyccel does not print the `if` block, causing a Fortran runtime error when running the executable:
```
$ ./array_creation_in_loop

At line 11 of file /Users/yaman/Desktop/issue-359/__pyccel__/array_creation_in_loop.f90
Fortran runtime error: Attempting to allocate already allocated variable 'a'

Error termination. Backtrace:
#0  0x10c4e7b6d
#1  0x10c4e8825
#2  0x10c4e8e06
#3  0x10c4dedd9
```

**Comment by yguclu** [comment]
Third minimal example: **array creation in if statement**. Program
```python
import numpy as np

i = 1
j = 2

if i == j:
    a = np.ones(5, int)
else:
    a = np.ones(7, int)
```
should simply be translated to
```fortran
program prog_array_creation_in_if

implicit none

integer(kind=8) :: i
integer(kind=8) :: j
integer(kind=8), allocatable :: a(:)

i = 1_8
j = 2_8
if (i == j) then
  allocate(a(0:5_8 - 1_8))
else
  allocate(a(0:7_8 - 1_8))
end if

end program prog_array_creation_in_if
```
provided that Pyccel can check if `dtype` and `rank` are the same in the two definitions. As a first step we can handle this case as in the first example (**array redefinition**) and then improve it as a further optimization.
